lib/axon/compiler.ex

```diff
@@ -225,6 +265,30 @@ defmodule Axon.Compiler do
     Nx.as_type(value, Nx.type(template))
   end
 
+  defp normalize_blocks(params) do
+    # Blocks are kinda hacky and produce a model state,
+    # so we normalize them so we get proper model metadata
+    # and then have just one root-level model state struct
+    {data, parameters, state} =
+      Enum.reduce(params, {%{}, %{}, %{}}, fn
+        {key, %Axon.ModelState{} = model_state}, {acc, parameters, state} ->
+          updated_parmeters =
+            if model_state.parameters == %{},
+              do: parameters,
+              else: Map.put(parameters, key, model_state.parameters)
+
+          updated_state =
+            if model_state.state == %{}, do: state, else: Map.put(state, key, model_state.state)
+
+          {Map.put(acc, key, model_state.data), updated_parmeters, updated_state}
+
+        {key, data}, {acc, parameters, state} ->
+          {Map.put(acc, key, data), parameters, state}
+      end)
+
+    {data, %{parameters: parameters, state: state}}
```

Comment:
So what are blocks? Are those ModelState inside ModelStates? Are they the primary motivation for adding ModelState?

Reply:
I would not add a map here:

```suggestion
    {data, parameters, state}
```

Otherwise we are putting it in a map only to access individual members later.